### PR TITLE
fix and then specialize sparse broadcast!(f, C, A), strengthen tests

### PR DESCRIPTION
**Pull request by Sacha0**
In https://github.com/JuliaLang/julia/pull/19895#issuecomment-271230830, @martinholters (thanks!) noticed that the existing sparse `broadcast!(f, C, A)` specialization does not handle cases where the shapes of destination `C` and source `A` differ.

This pull request's first commit removes the incorrect sparse `broadcast!(f, C, A)` specialization (falling back to the argument-number-generic sparse `broadcast!` code, which does not suffer from the above issue) and strengthens associated tests. The argument-number-generic code leaves some performance on the table, so the second commit provides a corrected sparse `broadcast!(f, C, A)` specialization for zero-preserving `f`, and the third commit likewise for not-zero-preserving `f`.

While writing this pull request I realized that the sparse `broadcast!(f, C, A, B)` specializations suffer from the same issue. A fix for that code is in the works. Best!

### CODE REVIEW

File: test/sparse/higherorderfns.jl [hunk 1]
```diff
@@ -76,20 +75,77 @@ end
...
     N, M, p = 10, 12, 0.4
     a, A = sprand(N, p), sprand(N, M, p)
     fa, fA = Array(a), Array(A)
     @test broadcast(sin, a) == sparse(broadcast(sin, fa))
     @test broadcast(sin, A) == sparse(broadcast(sin, fA))
-    @test broadcast!(sin, copy(a), a) == sparse(broadcast!(sin, copy(fa), fa))
-    @test broadcast!(sin, copy(A), A) == sparse(broadcast!(sin, copy(fA), fA))
+    # The two following tests are from #19895, and check the absence of an ambiguity
+    # between a pair of specializations, but #19895 removed both of those specializations,
+    # so perhaps these tests should be nixed?
```
Comment (Sacha0) [inline test/sparse/higherorderfns.jl]: @martinholters @pabloferz, thoughts? Thanks!

File: base/sparse/higherorderfns.jl [hunk 2]
```diff
@@ -408,7 +409,87 @@ end
...
+    isempty(C) && return _finishempty!(C)
+    spaceC::Int = min(length(storedinds(C)), length(storedvals(C)))
+    # C and A cannot have the same shape, as we directed that case to map in broadcast's
+    # entry point; here we need efficiently handle only heterogeneous C-A combinations where
+    # one or both of C and A has at least one singleton dimension.
+    #
+    # We first divide the cases into two groups: those in which the input argument does not
+    # expand vertically, and those in which the input argument expands vertically.
+    #
+    # Cases without version expansion
```
Comment (tkelman) [inline base/sparse/higherorderfns.jl]: vertical?

File: base/sparse/higherorderfns.jl [hunk 3]
```diff
@@ -408,7 +409,87 @@ end
...
+    else # numrows(A) != numrows(C) (=> numrows(A) == 1)
+        @inbounds for j in columns(C)
+            setcolptr!(C, j, Ck)
+            Ak, stopAk = numcols(A) == 1 ? (colstartind(A, 1), colboundind(A, 1)) : (colstartind(A, j), colboundind(A, j))
+            Ax = Ak < stopAk ? storedvals(A)[Ak] : zero(eltype(A))
+            fofAx = f(Ax)
+            # if fofAx is zero, then either A's jth column is empty, or A's jth column
+            # contains a nonzero value x but f(Ax) is nonetheless zero, so we need store
+            # nothing in C's jth column. if to the contrary fofAx is nonzero, then we must
+            # densely populate C's jth column wiht fofAx.
```
Comment (tkelman) [inline base/sparse/higherorderfns.jl]: with misspelled

File: base/sparse/higherorderfns.jl [hunk 4]
```diff
@@ -408,7 +409,87 @@ end
...
+    trimstorage!(C, Ck - 1)
+    return C
+end
+function _broadcast_notzeropres!{Tf}(f::Tf, fillvalue, C::SparseVecOrMat, A::SparseVecOrMat)
+    # For information on this code, see comments in similar code in _broadcast_zeropres! above
+    # Build dense matrix structure in C, expanding storage if necessary
+    _densestructure!(C)
+    # Populate values
+    fill!(storedvals(C), fillvalue)
+    # Cases without version expansion
```
Comment (tkelman) [inline base/sparse/higherorderfns.jl]: vertical

File: test/sparse/higherorderfns.jl [hunk 5]
```diff
@@ -76,20 +75,77 @@ end
     end
 end
 
-@testset "broadcast[!] implementation specialized for a single (input) sparse vector/matrix" begin
-    # broadcast[!] for a single sparse vector/matrix falls back to map[!], tested extensively
-    # above. here we simply lightly exercise the relevant broadcast[!] entry points.
+@testset "broadcast implementation specialized for a single (input) sparse vector/matrix" begin
+    # broadcast for a single (input) sparse vector/matrix falls back to map, tested
+    # extensively above. hre we simply lightly exercise the relevant broadcast entry
```
Comment (tkelman) [inline test/sparse/higherorderfns.jl]: "here" misspelled

File: test/sparse/higherorderfns.jl (hunk 1 above)
Comment (martinholters) [inline test/sparse/higherorderfns.jl]: A few months down the road, someone decides to add specializations for the common  `x .= y` case to be faster...

File: base/sparse/higherorderfns.jl (hunk 2 above)
Comment (Sacha0) [inline base/sparse/higherorderfns.jl]: Good catch. Fixed on push. Thanks!

File: base/sparse/higherorderfns.jl (hunk 3 above)
Comment (Sacha0) [inline base/sparse/higherorderfns.jl]: Yesterday was not a good day for spelling. Fixed on push. Thanks!

File: base/sparse/higherorderfns.jl (hunk 4 above)
Comment (Sacha0) [inline base/sparse/higherorderfns.jl]: Fixed on push. Thanks!

File: test/sparse/higherorderfns.jl (hunk 5 above)
Comment (Sacha0) [inline test/sparse/higherorderfns.jl]: Fixed on push. Thanks!

File: test/sparse/higherorderfns.jl (hunk 1 above)
Comment (Sacha0) [inline test/sparse/higherorderfns.jl]: On that note, perhaps we should add tests for https://github.com/JuliaLang/julia/pull/19895#issuecomment-272237030. Will add a commit. Thanks!

**Comment by Sacha0** [comment]
Added a commit strengthening `broadcast!(identity, C, A)` tests to guard against reintroduction of ambiguous and/or erroneous specializations. (Ref. #19895 / https://github.com/JuliaLang/julia/pull/19895#issuecomment-272237030). Best!

**Comment by Sacha0** [comment]
Absent objections or requests for time, I plan to merge this Thursday morning PST. Best!

**Comment by pabloferz** [comment]
LGTM

**Comment by Sacha0** [comment]
Thanks for reviewing / merging!

**Pull request merged**